llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -2041,6 +2041,25 @@ void CombinerHelper::applyCombineMulToShl(MachineInstr &MI,
   Observer.changedInstr(MI);
 }
 
+bool CombinerHelper::matchCombineSubToAdd(MachineInstr &MI,
+                                          BuildFnTy &MatchInfo) {
+  GSub &Sub = cast<GSub>(MI);
+
+  LLT Ty = MRI.getType(Sub.getReg(0));
+
+  APInt Imm = getIConstantFromReg(Sub.getRHSReg(), MRI);
+
```

Comment:
```
 if (!isLegalOrBeforeLegalizer({TargetOpcode::G_ADD, {Ty}}))
    return false;
```

Reply:
Feels very unlikely that any target would have G_SUB and not G_ADD.

Reply:
```
if (isConstantLegalOrBeforeLegalizer(Ty)) 
  return false;
```

Reply:
That's no excuse. We must always check for legality.

Reply:
We don't check legality in the mul to shift combine right above this code. So we don't always check.

Reply:
We always check for legality. GlobalISel is for me for weird devices. The combine above is buggy. Implied legality is a no go.